sandy/fy.py

```diff
@@ -445,24 +438,21 @@ def gls_update(self, zam, e, decay_data, xp, Vy=None, kind='cumulative',
 
         """
         data = self.data
-        S = _gls_setup(decay_data, kind, index_columns)
         # Filter FY data:
         fy_data = self.filter_by('ZAM', zam)\
                       .filter_by("E", e).data\
                       .set_index('ZAP')[['MT', 'FY', 'DFY']]
         # Divide the data:
         fy_type = fy_data.groupby('MT')
         IFY, CFY = [fy_type.get_group(x)['FY'] for x in fy_type.groups]
-        DIFY, DCFY = [fy_type.get_group(x)['DFY'] for x in fy_type.groups]
+        mask = (data.ZAM == zam) & (data.MT == 454) & (data.E == e)
+        DIFY = data.loc[mask, 'DFY']
+        # Fix the S with the correct dimension:
+        S = _gls_setup(decay_data, kind, index=IFY.index, columns=IFY.index)
         # Put the data in a appropiate format:
         DIFY = sandy.CategoryCov.from_var(DIFY).data
-        if Vy is None:
-            Vy = sandy.CategoryCov.from_var(DCFY).data
-        # Fix the index:
-        S = S.loc[IFY.index, IFY.index].fillna(0)
         # Perform GLS:
-        IFY_new = sandy.gls_update(IFY, S, DIFY, Vy, CFY, xp, threshold=threshold)
-        mask = (data.ZAM == zam) & (data.MT == 454) & (data.E == e)
+        IFY_new = sandy.gls_update(IFY, S, DIFY, Vy, CFY, y, threshold=threshold)
```

Comment:
I see here that there is something wrong with `sandy.gls_update`...`CFY` shouldn't be given. It should be internally calculated by as a solution of the linearized model `y=S x`